pysecs/secs.py

```diff
@@ -140,14 +140,24 @@ def fit(
                 # Find singular value decompostion
                 U, S, Vh = np.linalg.svd(svd_in, full_matrices=False)
 
-                # Eliminate singular values less than epsilon by setting their
-                # reciprocal to zero (setting S to infinity firsts avoids
-                #  divide-by-zero warings)
-                S[S < epsilon * S.max()] = np.inf
+                # Compute cumulative energy (variance explained)
+                cumulative_energy = np.cumsum(S**2)
+                total_energy = cumulative_energy[-1]
+
+                # Find how many components retain 95% of the total energy
+                energy_ratio = cumulative_energy / total_energy
+                # +1 to get the index after insertion
+                n_components = np.searchsorted(energy_ratio, 1 - epsilon) + 1
```

Comment:
I believe most folks that use SVD think of epsilon as a fraction of a singular value, which is the square-root of a variance. If you agree, shouldn't the criterion then be `1 - epsilon**2`?

Or do you want to encourage folks to think of epsilong as a fractional variance?

I don't have a preference, but just wanted to be certain you considered this.

Reply:
I wonder if we need a different keyword for this case then? I don't think it should be epsilon**2, as I really think that what we are doing here is keeping a number of components that satisfy an arbitrary (1-epsilon here) fraction/percentage of variance. I agree it is fundamentally different than keeping values that are within 95% of the maximum component's standard deviation.

See numbers 7-8 here where it is talking about sum of energy as a percentage and keeping the top 90% for components. (Maybe we don't need to square "S" when doing this though?)
https://en.wikipedia.org/wiki/Principal_component_analysis#Computation_using_the_covariance_method

Reply:
It's not completely obvious, but that wiki does actually recommend that you square "S". It awkwardly inserts SVD into the discussion, casually noting that singular values are the square root of the eigenvalues, then goes on to associate eigenvalues with "cumulative energy".

In any case, I think I'll change my non-recommendation for `1 - epsilon**2`, an suggest that you add a new option to override the default original behavior. This way there is not a breaking change (you do have a bit of a user base). You might call that new option "energy", although maybe that isn't the most intuitive.

I'll leave it to you (obviously), and you can consider this comment "resolved" as far as I'm concerned.

Reply:
I added a new argument "mode" to `fit()` here defaulting to the previous version. Let me know what you think of the naming/implementation with this new addition.

Reply:
I think this latest change looks great. A concise "mode" name that actually encompasses everything about the new method probably does not exist, but the updated doc string makes it perfectly clear what is happening.